Bjorncs/improve athenz access control

bjorncs opened the pull request:
Requires https://github.com/vespa-engine/vespa/pull/13930 to be merged.


FYI @gjoranv @tokle

bjorncs committed:
Bind access control filter to all paths on port 4443

Exclude bindings with separate filter chain.
Do not bind access control filter to port 4080.
Simplify AccessControl interface to a 'configure()' method.

bjorncs committed:
Use 'createModelAndGetHttp' in all test methods for AccessControl

bjorncs committed:
Implement equals/hashCode for FilterBinding

bjorncs committed:
Ensure that access control chain has unique bindings

bjorncs committed:
Add static builder taking port and path

gjoranv commented on config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java:
```diff
@@ -137,6 +139,22 @@ private void addAccessControlExcludedChain(Http http) {
+    private void removeDuplicateBindingsFromAccessControlChain(Http http) {
+        Set<FilterBinding> duplicateBindings = new HashSet<>();
+        for (FilterBinding binding : http.getBindings()) {
+            if (binding.filterId().toId().equals(ACCESS_CONTROL_CHAIN_ID)) {
```
BTW, is `filterId()` the correct name? Sounds like it should be `chainId`.

gjoranv reviewed (commented):
It's great that you removed the code duplication in the builders. Otherwise, please see comments.

gjoranv commented:
Build fails with:
```
java.lang.IllegalArgumentException: Could not create a component with id 'com.yahoo.jdisc.http.filter.security.misc.NoopFilter'. Tried to load class directly, since no bundle was found for spec: jdisc-security-filters. If a bundle with the same name is installed, there is a either a version mismatch or the installed bundle's version contains a qualifier string.
```

bjorncs committed:
Delegate configuration of noop filter to amender

The open-source model cannot configure the the noop filter, otherwise unit tests using application framework will fail when 'access-control' is in services.xml.

bjorncs commented:
> Build fails with:
> 
> ```
> java.lang.IllegalArgumentException: Could not create a component with id 'com.yahoo.jdisc.http.filter.security.misc.NoopFilter'. Tried to load class directly, since no bundle was found for spec: jdisc-security-filters. If a bundle with the same name is installed, there is a either a version mismatch or the installed bundle's version contains a qualifier string.
> ```

This is fixed with https://github.com/vespa-engine/vespa/pull/13942/commits/8373e8d51ff3019e94f82211df16efae23080fe6

bjorncs commented on config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java:
```diff
@@ -84,69 +85,75 @@ public AccessControl build() {
-    public List<FilterBinding> getBindings() {
-        return Stream.concat(getHandlerBindings(), getServletBindings())
-                .collect(Collectors.toCollection(ArrayList::new));
+    public void configure(Http http) {
```
Perhaps rename it to something like `configureAccessControlFilterChains`?

I would like to contain the logic for setting up filter chains and bindings inside the `AccessControl` class.

gjoranv replied:
Isn't `configureAccessControl` enough? I'm ok with with either. I agree with containing the logic.

bjorncs replied:
Renamed to `configureHttpFilterChains`.

bjorncs commented on config-model/src/main/java/com/yahoo/vespa/model/container/component/SystemBindingPattern.java:
```diff
@@ -13,6 +13,7 @@
 
     public static SystemBindingPattern fromHttpPath(String path) { return new SystemBindingPattern("http", "*", null, path);}
     public static SystemBindingPattern fromPattern(String binding) { return new SystemBindingPattern(binding);}
+    public static SystemBindingPattern fromPortAndPath(String port, String path) { return new SystemBindingPattern("http", "*", port, path); }
```
I agree, let's rename it to `fromHttpPortAndPath`.

bjorncs committed:
Rename 'fromPortAndPath' -> 'fromHttpPortAndPath'

bjorncs committed:
Don't duplicate container element in all unit tests

bjorncs committed:
Component spec in FilterBinding refers to a chain not filter

bjorncs commented on config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java:
```diff
@@ -137,6 +139,22 @@ private void addAccessControlExcludedChain(Http http) {
+    // Remove bindings from access control chain that have binding pattern as a different filter chain
+    private void removeDuplicateBindingsFromAccessControlChain(Http http) {
+        Set<FilterBinding> duplicateBindings = new HashSet<>();
+        for (FilterBinding binding : http.getBindings()) {
```
There are multiple strategies to solve this specific problem, but I don't think creating a mapping will be significantly better.

bjorncs committed:
Rename 'configure' -> 'configureHttpFilterChains'

bjorncs commented on config-model/src/test/java/com/yahoo/vespa/model/container/xml/AccessControlTest.java:
```diff
@@ -44,15 +42,14 @@
     public void access_control_filter_chains_are_set_up() {
-        Element clusterElem = DomBuilderTest.parse(
+        Http http = createModelAndGetHttp(
+                "<container version='1.0'>",
```
Fixed

bjorncs commented:
@gjoranv PTAL

gjoranv reviewed (approved):
👍
